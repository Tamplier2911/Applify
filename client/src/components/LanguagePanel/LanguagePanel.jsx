import "./LanguagePanel.scss";
import React from "react";

// redux
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";
import {
  setLanguageEng,
  setLanguageRus,
  setLanguageUkr,
  toggleLanguagePanel
} from "../../redux/lang/lang.actions";
import {
  selectCurrentLanguage,
  selectLanguagePanelHidden
} from "../../redux/lang/lang.selectors";

// images
import langEng from "../../assets/jpeg/lang-eng.jpg";
import langUkr from "../../assets/jpeg/lang-ukr.jpg";
import langRus from "../../assets/jpeg/lang-rus.jpg";

// svg
import { ReactComponent as LangPannelSVG } from "../../assets/svg/language.svg";

const LanguagePanel = ({
  lang,
  panelHidden,
  setLanguageEng,
  setLanguageRus,
  setLanguageUkr,
  toggleLanguagePanel
}) => {
  return (
    <div className="language-panel">
      {panelHidden ? null : (
        <div className="language-panel__languages">
          <div
            className="language-panel__language-1"
            onClick={() => setLanguageEng()}
          >
            <img
              className="language-panel__image"
              alt="great britain flag"
              src={langEng}
            ></img>
          </div>
          <div
            className="language-panel__language-2"
            onClick={() => setLanguageUkr()}
          >
            <img
              className="language-panel__image"
              alt="ukrainian flag"
              src={langUkr}
            ></img>
          </div>
          <div
            className="language-panel__language-3"
            onClick={() => setLanguageRus()}
          >
            <img
              className="language-panel__image"
              alt="russian flag"
              src={langRus}
            ></img>
          </div>
        </div>
      )}

      <div
        className="language-panel__wrapper"
        onClick={() => toggleLanguagePanel()}
      >
        <LangPannelSVG className="language-panel__svg">
          Language Pannel
        </LangPannelSVG>
      </div>
    </div>
  );
};

const mapStateToProps = createStructuredSelector({
  lang: selectCurrentLanguage,
  panelHidden: selectLanguagePanelHidden
});

export default connect(mapStateToProps, {
  setLanguageEng,
  setLanguageRus,
  setLanguageUkr,
  toggleLanguagePanel
})(LanguagePanel);
