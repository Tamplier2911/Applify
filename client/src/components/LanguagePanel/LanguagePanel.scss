/*
.language-panel {
  position: fixed;
  bottom: 4rem;
  right: 4rem;
  z-index: 20;

  &__language-1 {
    cursor: pointer;
    position: absolute;
    // top: 2rem;
    // left: 2rem;
    top: -5rem;
    left: 1.8rem;

    animation: engAnimation 0.5s;

    width: 4rem;
    height: 4rem;
    border-radius: 50%;
    overflow: hidden;
    box-shadow: 0.1rem 0.2rem 0.6rem rgba(0, 0, 0, 0.623);
  }

  &__language-2 {
    cursor: pointer;
    position: absolute;
    // top: 2rem;
    // left: 2rem;
    top: -2.4rem;
    left: -4.2rem;

    animation: rusAnimation 0.5s;

    width: 4rem;
    height: 4rem;
    border-radius: 50%;
    overflow: hidden;
    box-shadow: 0.1rem 0.2rem 0.6rem rgba(0, 0, 0, 0.623);
  }

  &__language-3 {
    cursor: pointer;
    position: absolute;
    // top: 2rem;
    // left: 2rem;
    top: 4rem;
    left: -4.8rem;

    animation: ukrAnimation 0.5s;

    width: 4rem;
    height: 4rem;
    border-radius: 50%;
    overflow: hidden;
    box-shadow: 0.1rem 0.2rem 0.6rem rgba(0, 0, 0, 0.623);
  }

  &__image {
    width: 100%;
    height: 100%;
    display: block;
    object-fit: cover;
  }

  &__wrapper {
    position: relative;
    cursor: pointer;
    display: flex;
    align-items: center;
    justify-content: center;
    background-color: #fff;
    box-shadow: 0.1rem 0.2rem 0.6rem rgba(0, 0, 0, 0.623);
    border-radius: 50%;
    width: 7rem;
    height: 7rem;
    transition: transform 0.3s;

    &:hover {
      transform: scale(1.1);
    }

    &:hover .language-panel__svg {
      fill: var(--cl-primary);
    }
  }

  &__svg {
    fill: var(--cl-font);
    width: 4rem;
    height: 4rem;
    transition: fill 0.3s;

    &:hover {
      fill: var(--cl-primary);
    }
  }
}

@keyframes engAnimation {
  0% {
    top: 2rem;
    left: 2rem;
  }
  100% {
    top: -5rem;
    left: 1.8rem;
  }
}

@keyframes rusAnimation {
  0% {
    top: 2rem;
    left: 2rem;
  }
  100% {
    top: -2.4rem;
    left: -4.2rem;
  }
}

@keyframes ukrAnimation {
  0% {
    top: 2rem;
    left: 2rem;
  }
  100% {
    top: 4rem;
    left: -4.8rem;
  }
}
*/
