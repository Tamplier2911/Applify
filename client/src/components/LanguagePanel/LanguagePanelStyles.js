import styled, { css } from "styled-components";

import { ReactComponent as LangPannelSVG } from "../../assets/svg/language.svg";

const languageCommonStyles = css`
  cursor: pointer;
  position: absolute;
  width: 4rem;
  height: 4rem;
  border-radius: 50%;
  overflow: hidden;
  box-shadow: 0.1rem 0.2rem 0.6rem var(--cl-langpan-shadow);
`;

const engAnimation = css`
  @keyframes engAnimation {
    0% {
      top: 2rem;
      left: 2rem;
    }
    100% {
      top: -5rem;
      left: 1.8rem;
    }
  }
`;

const rusAnimation = css`
  @keyframes rusAnimation {
    0% {
      top: 2rem;
      left: 2rem;
    }
    100% {
      top: -2.4rem;
      left: -4.2rem;
    }
  }
`;

const ukrAnimation = css`
  @keyframes ukrAnimation {
    0% {
      top: 2rem;
      left: 2rem;
    }
    100% {
      top: 4rem;
      left: -4.8rem;
    }
  }
`;

export const LanguagePanelContainer = styled.div`
  position: fixed;
  bottom: 4rem;
  right: 4rem;
  z-index: 20;
`;

export const LanguagePanelLanguages = styled.div``;

export const LanguagePanelLangEng = styled.div`
  top: -5rem;
  left: 1.8rem;
  animation: engAnimation 0.5s;
  ${languageCommonStyles}
  ${engAnimation}
`;

export const LanguagePanelLangRus = styled.div`
  top: -2.4rem;
  left: -4.2rem;
  animation: rusAnimation 0.5s;
  ${languageCommonStyles}
  ${rusAnimation}
`;

export const LanguagePanelLangUkr = styled.div`
  top: 4rem;
  left: -4.8rem;
  animation: ukrAnimation 0.5s;
  ${languageCommonStyles}
  ${ukrAnimation}
`;

export const LanguagePanelImage = styled.img`
  width: 100%;
  height: 100%;
  display: block;
  object-fit: cover;
`;

export const LanguagePanelWrap = styled.div`
  position: relative;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: var(--cl-white);
  box-shadow: 0.1rem 0.2rem 0.6rem var(--cl-langpan-shadow);
  border-radius: 50%;
  width: 7rem;
  height: 7rem;
  transition: transform 0.3s;

  &:hover {
    transform: scale(1.1);
  }

  &:hover .language-panel__svg {
    fill: var(--cl-primary);
  }
`;

export const LanguagePanelSVG = styled(LangPannelSVG)`
  fill: var(--cl-font);
  width: 4rem;
  height: 4rem;
  transition: fill 0.3s;

  &:hover {
    fill: var(--cl-primary);
  }
`;
