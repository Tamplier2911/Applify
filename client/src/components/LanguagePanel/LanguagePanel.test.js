import React from "react";

// component
import LanguagePanel from "./LanguagePanel";

// utils
import {
  setShallowC,
  findByTestAttr,
  storeFactory,
} from "../../tests/testUtils";

describe("<LanguagePanel />", () => {
  const initialStore = {
    lang: {
      languagePanelHidden: false,
      currentLanguage: "eng",
    },
  };
  const store = storeFactory(initialStore);

  let wrapper;
  beforeEach(() => {
    wrapper = setShallowC(<LanguagePanel store={store} />);
  });

  it("renders without an error", () => {
    const languagePanel = findByTestAttr(wrapper, "language-panel");
    expect(languagePanel.length).toBe(1);
  });

  it("do not render langauges if pannel is hidden", () => {
    const initialStore = {
      lang: {
        languagePanelHidden: true,
        currentLanguage: "eng",
      },
    };
    const store = storeFactory(initialStore);
    wrapper = setShallowC(<LanguagePanel store={store} />);
    const languagePanelLanguages = findByTestAttr(
      wrapper,
      "language-panel-languages"
    );
    expect(languagePanelLanguages.length).toBe(0);
  });

  it("set currentLanguage to 'eng' when setLanguageEng is clicked", () => {
    const languagePanelLangEng = findByTestAttr(
      wrapper,
      "language-panel-lang-eng"
    );
    expect(languagePanelLangEng.length).toBe(1);
    // perform click on setLanguageEng
    languagePanelLangEng.simulate("click");
    const {
      lang: { currentLanguage, languagePanelHidden },
    } = store.getState();
    expect(currentLanguage).toBe("eng");
    expect(languagePanelHidden).toBeTruthy();
    // turn "ON" language panel by clicking toggleLanguagePanel
    const toggleLanguagePanel = findByTestAttr(
      wrapper,
      "language-panel-toggle-btn"
    );
    toggleLanguagePanel.simulate("click");
  });

  it("set currentLanguage to 'ukr' when setLanguageUkr is clicked", () => {
    const languagePanelLangUkr = findByTestAttr(
      wrapper,
      "language-panel-lang-ukr"
    );
    expect(languagePanelLangUkr.length).toBe(1);
    // perform click on setLanguageUkr
    languagePanelLangUkr.simulate("click");
    const {
      lang: { currentLanguage },
    } = store.getState();
    expect(currentLanguage).toBe("ukr");
    // turn "ON" language panel by clicking toggleLanguagePanel
    const toggleLanguagePanel = findByTestAttr(
      wrapper,
      "language-panel-toggle-btn"
    );
    toggleLanguagePanel.simulate("click");
  });

  it("set currentLanguage to 'rus' when setLanguageRus is clicked", () => {
    const languagePanelRus = findByTestAttr(wrapper, "language-panel-lang-rus");
    expect(languagePanelRus.length).toBe(1);
    // perform click on setLanguageRus
    languagePanelRus.simulate("click");
    const {
      lang: { currentLanguage },
    } = store.getState();
    expect(currentLanguage).toBe("rus");
    // turn "ON" language panel by clicking toggleLanguagePanel
    const toggleLanguagePanel = findByTestAttr(
      wrapper,
      "language-panel-toggle-btn"
    );
    toggleLanguagePanel.simulate("click");
  });

  it("toggle language pannel when toggleLanguagePanel is clicked", () => {
    // assert on toggling
    const toggleLanguagePanel = findByTestAttr(
      wrapper,
      "language-panel-toggle-btn"
    );
    toggleLanguagePanel.simulate("click");
    let {
      lang: { languagePanelHidden },
    } = store.getState();
    expect(languagePanelHidden).toBeTruthy();
    toggleLanguagePanel.simulate("click");
    ({
      lang: { languagePanelHidden },
    } = store.getState());
    expect(languagePanelHidden).toBeFalsy();
  });

  it("matches snapshot", () => {
    const languagePanel = findByTestAttr(wrapper, "language-panel");
    expect(languagePanel).toMatchSnapshot();
  });
});
