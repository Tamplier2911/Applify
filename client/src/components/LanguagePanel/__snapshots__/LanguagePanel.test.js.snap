// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LanguagePanel /> matches snapshot 1`] = `
<styled.div
  data-test="language-panel"
>
  <styled.div
    data-test="language-panel-languages"
  >
    <styled.div
      data-test="language-panel-lang-eng"
      onClick={[Function]}
    >
      <styled.img
        alt="great britain flag"
        src="lang-eng.jpg"
      />
    </styled.div>
    <styled.div
      data-test="language-panel-lang-ukr"
      onClick={[Function]}
    >
      <styled.img
        alt="ukrainian flag"
        src="lang-ukr.jpg"
      />
    </styled.div>
    <styled.div
      data-test="language-panel-lang-rus"
      onClick={[Function]}
    >
      <styled.img
        alt="russian flag"
        src="lang-rus.jpg"
      />
    </styled.div>
  </styled.div>
  <styled.div
    data-test="language-panel-toggle-btn"
    onClick={[Function]}
  >
    <Styled(Component)>
      Language Pannel
    </Styled(Component)>
  </styled.div>
</styled.div>
`;
